### PR TITLE
Convert open/fetch to stream

### PR DESCRIPTION
**Pull request by arashout**
Problem:
Right now as seen from issue: https://github.com/nushell/nushell/issues/1448 it hangs when trying to open a file like /dev/random

Summary:
- Convert `open` command to work via `Streams` to allow `CTRL-C`ing mid-process,
- Move `MaybeTextCodec` since it was getting large
- There is a `fetch` function within open which is used for `enter`, I left this as a Future but did not convert it to a stream
- Reduced some code duplication within `open` e.g. `encoding_rs` already does BOM sniffing and decoding to utf-16 automatically 
   - Instead of doing custom BOM sniffing, I instead use the presence/absence of replacement characters to decide if we should use Binary or not
- Tried to add comments to many places

With this change I've broken the auto-convert functionality (Since it does not work with streams) of `open`, and I'm currently trying to determine what would be the best way to add it back in.

**Possible Methods**

1. Look at the extension of the file, if it is a recognizable one that we can auto-convert then collect the `Stream` and try to auto-convert **(Current implementation)**
2. Somehow convert AutoConvert to accept Streams
3. ???

However, the bug mentioned above is addressed, now `nu` no longer hangs with large files / streaming files with this streaming approach.

Demo:
https://www.youtube.com/watch?v=gDPCD5bqi9A&feature=youtu.be

TODO:
- [x] Convert `open` to `Stream` (Get types to line up)
- [x] Implement decoding logic for utf-8, utf-16, etc for `MaybeTextCodec` so we can decode non-utf-8 streams 
- [ ] Make `AutoConvert` accept streams?

### CODE REVIEW

**Comment by fdncred** [comment]
Nice video! It's so much easier to communicate functionality this way.  What about `bat` autoview? Does that still work? For instance, when I do `open src/main.rs` I get this today.
image

**Comment by arashout** [comment]
huh, that's interesting I would have thought that, that would have worked as a well with my workaround, but it doesnt

**Comment by jonathandturner** [comment]
Yeah, textview assumes the content is a single Value of all the text, tagged with the original location it came from. If you're outputting more than one, yeah, we'll have to update textview, too

**Comment by arashout** [comment]
I'm going to add the workaround for `bat` then as well.

**Comment by arashout** [comment]
Kind of weird but the code that it is erroring on seems like old code?

**Comment by jonathandturner** [comment]
@arashout - Looks like the shape of `MaybeTextCodec` now has an encoding field.  Anywhere it's created will need to be updated to add an encoding, too.

**Comment by arashout** [comment]
ah, I've broken TextView (bat integration) again after merging master

File: crates/nu-cli/src/commands/constants.rs [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    "xsd",
+    "xslt",
+    "tld",
+    "dtml",
+    "rss",
+    "opml",
+    "svg",
+    "yaml",
+    "yml",
+    "sublime-syntax",
```
Comment (jonathandturner) [inline crates/nu-cli/src/commands/constants.rs:357]: I wonder if we want to make a follow-up PR to get this list from bat. If that's not possible, we might want to file an issue with bat to make it possible to query the list from their library.

**Comment by jonathandturner** [comment]
Great job! Looks good, and everything I tried worked great.

**Pull request merged**